### PR TITLE
adaptor: Fix problem in LoadEnv in cloud Providers

### PR DESCRIPTION
**Pull request by huoqifeng**
Fixes: #1454

This PR will fix the `Fix the LoadEnv() function` described in #1454

### CODE REVIEW

**Review by surajssd — COMMENTED** [review]
Azure code looks good. Now I wonder if CAA really needs flags? If we extensively rely on using env vars?

File: pkg/adaptor/cloud/aws/manager.go [hunk 1]
```diff
@@ -38,9 +39,40 @@ func (_ *Manager) ParseCmd(flags *flag.FlagSet) {
 }
 
 func (_ *Manager) LoadEnv() {
+	// overwrite config set by cmd parameters in oci image with env might come from orchastration platform
 	cloud.DefaultToEnv(&awscfg.AccessKeyId, "AWS_ACCESS_KEY_ID", "")
 	cloud.DefaultToEnv(&awscfg.SecretKey, "AWS_SECRET_ACCESS_KEY", "")
 
+	cloud.DefaultToEnv(&awscfg.Region, "AWS_REGION", "")
```
[the remark this answers is not in the document]
Comment (huoqifeng) [reply pkg/adaptor/cloud/aws/manager.go:46]: @bpradipt envs aligned with the list in [entrypoint.sh](https://github.com/confidential-containers/cloud-api-adaptor/blob/main/entrypoint.sh.), but yes, we can keep a smaller list. It depends on providers.

**Review by genjuro214 — APPROVED** [review]
LGTM for ibmcloud changes

File: pkg/adaptor/cloud/aws/manager.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (huoqifeng) [reply pkg/adaptor/cloud/aws/manager.go:46]: @bpradipt is there any envs that you want drop in current change? I can revise the aws one accordingly.

**Comment by huoqifeng** [comment]
> Azure code looks good. Now I wonder if CAA really needs flags? If we extensively rely on using env vars?

Let's talk about it more on this weeks meeting and change it in further PR if we decided flags is not necessary? @bpradipt @surajssd , I want this get fixed soon so that we can enable the caa upgrade test cases.

**Review by liudalibj — APPROVED** [review]
LGTM

**Review by bpradipt — APPROVED** [review]
/lgtm
Would also request @snir911 to have a look once

**Comment by huoqifeng** [comment]
@surajssd do you mind approve this PR as we agreed to handle the flags separately?

**Comment by huoqifeng** [comment]
> /lgtm Would also request @snir911 to have a look once

@snir911 any comments on this?

**Comment by snir911** [comment]
Sorry @huoqifeng, i missed this PR
This PR should be fine, however, AFAIR it's not required from all variables to be loaded by LoadEnv, it's required only for variables which are needed for the provider's DeleteInstance implementation to work properly (for example for AWS the AWS_SECRET and AWS_ACCESS_KEY should be enough to execute AWS's [DeleteInstance](https://github.com/confidential-containers/cloud-api-adaptor/blob/dcb2cb74cb73abe2d2d9a7f3dc1a33097685ab77/pkg/adaptor/cloud/aws/provider.go#L315)  .
IIRC it was tested with Libvirt, AWS and Azure, so for these cloud providers it should have work just fine already.. for the other providers some adaptions are probably indeed required..

**Review by surajssd — APPROVED** [review]

**Comment by huoqifeng** [comment]
Rollback the changes in aws, would you please check it @snir911 ?

**Comment by bpradipt** [comment]
> Sorry @huoqifeng, i missed this PR This PR should work fine, however, AFAIR it's not required from all variables to be loaded by LoadEnv, it's required only for variables which are needed for the provider's DeleteInstance implementation to work properly (for example for AWS the AWS_SECRET and AWS_ACCESS_KEY should be enough to execute AWS's [DeleteInstance](https://github.com/confidential-containers/cloud-api-adaptor/blob/dcb2cb74cb73abe2d2d9a7f3dc1a33097685ab77/pkg/adaptor/cloud/aws/provider.go#L315) . IIRC it was tested with Libvirt, AWS and Azure, so for these cloud providers it should have work just fine already.. for the other providers some adaptions are probably indeed required..

I was thinking setting only the required variables should be fine. I also don't see any harm in setting all the variables. Just that we need to ensure to keep the cli params and env variables in sync. 

Also we need to be consistent in the approach for all the providers. Otherwise it's a pain to maintain the code going forward.

Let me ask few related questions specifically to @huoqifeng and @snir911 (others are also welcome to add their thoughts :-) ) to help with deciding on the next steps :
1. Why should only the required variables be set ?
2. What are the benefits of loading all the config params as environment variables ?
3. What are the drawbacks of loading all the config params as environment variables ?

@huoqifeng sorry this PR is taking more time. But I'm hoping this discussion and eventual decision will be worth the time.

**Comment by snir911** [comment]
> 1. Why should only the required variables be set ?

No special reason AFAIR, just to avoid passing unnecessary parameters which will be unused

> 2. What are the benefits of loading all the config params as environment variables ?

I don's see much benefit in doing it for current code, however, i do see advantages if will do it for all parameters, for all providers and making sure all variables naming are comply with some set of rules but this would be quite a refactoring. I think this will allow to simplify code and move logic out from the entrypoint.sh.

> 3. What are the drawbacks of loading all the config params as environment variables ?

Not many i guess..
a. unnecessary parameters which will be passed to peerpod-ctrl
b. if someone is still running CAA as a service it will not be very convenient 
c. without doing what i mentioned above variables names are still inconsistent and confusing 
however, I'm not sure if this is best practice, nor if there are any security implications.

**Comment by snir911** [comment]
@huoqifeng thanks, I'd only drop the commit instead of reverting it.
BTW I'd also drop Azure unless you know it's required for it to be working

**Comment by huoqifeng** [comment]
Squashed the commits and keep aws and azure as @snir911 suggested. @snir911 I did not try azure, I'm happy to keep aws and azure as is if you already tested it. It's good to understand the purpose of the flags and LoadEnv. For any further enhancements, lets discuss in a separate issues. Thanks!

**Review by snir911 — APPROVED** [review]
thanks @huoqifeng !

**Pull request merged**